packages/backend/src/api/routers/FrontendRouter.ts

```diff
@@ -10,25 +16,51 @@ export function createFrontendRouter(frontendController: FrontendController) {
     ctx.body = await frontendController.getHomePage()
   })
 
-  router.get('/state-updates', async (ctx) => {
-    const page = parseInt(String(ctx.query.page ?? '1'))
-    const perPage = parseInt(String(ctx.query.perPage ?? '10'))
-    if ([page, perPage].some(Number.isNaN)) {
-      ctx.status = 500
-      return
-    }
-    ctx.body = await frontendController.getStateChangesPage(page, perPage)
-  })
+  router.get(
+    '/state-updates',
+    withTypedContext(
+      z.object({
+        query: z.object({
+          page: stringToPositiveInt('1'),
+          perPage: stringToPositiveInt('10'),
+        }),
+      }),
+      async (ctx) => {
+        const { page, perPage } = ctx.query
+        ctx.body = await frontendController.getStateChangesPage(page, perPage)
+      }
+    )
+  )
```

Comment:
I really liked your solution, I just had a thought on how can this actually be improved. I wasn't sure if changing the type of `ctx` can actually work, so I quickly changed the code to see if it was possible, but once I had the code I just decided that I can push it instead of making you implement the change again :)

Reply:
Let me know what do you think about this change.